searx/engines/google_news.py

```diff
@@ -16,74 +45,155 @@
-        if language:
-            params['url'] += '&hl=' + language
+    """Google-News search request"""
+
+    language, country, lang_country = get_lang_country(
+        # pylint: disable=undefined-variable
+        params, supported_languages, language_aliases
+    )
+    subdomain = 'news.google.com'
+
+    if params['time_range']: # in time_range_dict:
+        query += ' ' + time_range_dict[params['time_range']]
+
+    query_url = 'https://'+ subdomain + '/search' + "?" + urlencode({
+        'q': query,
+        'hl': language,
+        'lr': "lang_" + language,
+        'ie': "utf8",
+        'oe': "utf8",
+        'ceid' : "%s:%s" % (country, language),
+        'gl' : country,
+    })
+
+    if params['safesearch']:
+        query_url += '&' + urlencode({'safe': filter_mapping[params['safesearch']]})
+
+    params['url'] = query_url
+    logger.debug("query_url --> %s", query_url)
+
+    # en-US,en;q=0.8,en;q=0.5
+    params['headers']['Accept-Language'] = (
+        lang_country + ',' + language + ';q=0.8,' + language + ';q=0.5'
+        )
+    logger.debug("HTTP header Accept-Language --> %s",
+                 params['headers']['Accept-Language'])
+    params['headers']['Accept'] = (
+        'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8'
+        )
+    #params['google_subdomain'] = subdomain
 
     return params
 
 
-# get response from search-request
 def response(resp):
+    """Get response from google's search request"""
     results = []
 
+    # detect google sorry
+    resp_url = urlparse(resp.url)
+    if resp_url.netloc == 'sorry.google.com' or resp_url.path == '/sorry/IndexRedirect':
+        raise RuntimeWarning('sorry.google.com')
+
+    if resp_url.path.startswith('/sorry'):
+        raise RuntimeWarning(gettext('CAPTCHA required'))
+
+    # which subdomain ?
+    # subdomain = resp.search_params.get('google_subdomain')
+
+    # convert the text to dom
     dom = html.fromstring(resp.text)
 
-    # parse results
-    for result in dom.xpath('//div[@class="g"]|//div[@class="g _cy"]'):
+    for result in eval_xpath(dom, '//div[@class="xrnccd"]'):
+
         try:
-            r = {
-                'url': result.xpath('.//a[@class="l lLrAF"]')[0].attrib.get("href"),
-                'title': ''.join(result.xpath('.//a[@class="l lLrAF"]//text()')),
-                'content': ''.join(result.xpath('.//div[@class="st"]//text()')),
-            }
-        except:
+            # The first <a> tag in the <article> contains the link to the
+            # article The href attribute of the <a> is a google internal link,
+            # we can't use.  The real link is hidden in the jslog attribute:
+            #
+            #   <a ...
+            #      jslog="95014; 4:https://www.cnn.com/.../index.html; track:click"
+            #      href="./articles/CAIiENu3nGS...?hl=en-US&amp;gl=US&amp;ceid=US%3Aen"
+            #      ... />
+
+            jslog = eval_xpath(result, './article/a/@jslog')[0]
```

Comment:
This line can rewritten this way:
```python
title = extract_text(eval_xpath_getindex(result, './article/a/@jslog', 0))
```

eval_xpath_getindex [raises](https://github.com/searx/searx/blob/f310305c54c3cd1d9fc74f09453294edbd2b5486/searx/utils.py#L559-L609) an [SearxEngineXPathException](https://github.com/searx/searx/blob/f310305c54c3cd1d9fc74f09453294edbd2b5486/searx/exceptions.py#L96-L103) when the XPath is not found or when the index 0 is not found. This exception include the XPath.

Then this exception can [recorded](https://github.com/searx/searx/blob/f310305c54c3cd1d9fc74f09453294edbd2b5486/searx/search/processors/online.py#L155).

Then some global statistics can be seen at the [`/stats/errors` URL](https://github.com/searx/searx/blob/f310305c54c3cd1d9fc74f09453294edbd2b5486/searx/webapp.py#L952-L977)

But this doesn't work if there is global try ...  except. Is it possible to remove it ?

Reply:
Thanks a lot for the explanation .. I didn't aware of. I will remove try/except and make use of the new eval_xpath_[getindex|list]